ls1intum/Artemis
# `General`: Allow admins to perform a batch deletion of users

### Checklist
#### General


#### Client

### Motivation and Context



Closes https://github.com/ls1intum/Artemis/issues/5061: Allow admins to perform a batch deletion of users

### Description


We added a new endpoint to delete a list of user: `/users?login=...&login=...` - this endpoints returns the list of deleted users and the number of deleted users. The current user can't be deleted. Errors during deletion are logged and the user causing problems is skipped.

In addition, we added a new button to batch delete all selected users. The user needs to confirm this operation.
We added an additional column to allow selection of users.

Finally, we add server-side validation to prevent deletion of one's account.

In order to increase the test coverage, we added additional server and client tests.

### Steps for Testing

Prerequisites:
- 1 Admin
- Multiple Users/Accounts

1. Log in to Artemis
2. Navigate to User Management
3. Create 2+ new Users
4. Select those Users
5. Click the delete button
6. Verify that those users are deleted

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| user-management.component.ts | 82.35% | 94% |
| UserResource.java (newly added endpoint) | >90%* | >90%* |

\* flaky server tests (scheduleService), but newly added endpoint is covered with 3/4 tests (`UserBambooBitbucketJiraIntegrationTest` and `UserJenkinsGitlabIntegrationTest`), therefore the newly added endpoint is well covered.

### Screenshots



<details>
<summary>User Management</summary>

![image](https://user-images.githubusercontent.com/60005702/175261800-ff337f2f-7511-46e3-9d19-7f09e2c30627.png)
![image](https://user-images.githubusercontent.com/60005702/175261858-1f79aafe-d7c9-4bf1-96a9-2edc9ab92c8a.png)
![image](https://user-images.githubusercontent.com/60005702/175261921-4de7cbb1-3db1-4287-9be5-71e10f05f145.png)
![image](https://user-images.githubusercontent.com/60005702/175262098-79a20381-9c00-4871-a1b2-131cb435d9d3.png)
</details>
